### PR TITLE
Adding RAI attributes to the editor

### PR DESCRIPTION
**Pull request by JoanGi**
Following our discussion during the Croissant task force meeting, this pull request attempts to add the Croissant RAI Extension attributes to the editor. I've added a new Tab clearly marked “Extension: Responsible AI.” Where users can fill in the RAI attributes, this kind of "Tab extension" could serve as a guide for more extensions. For instance, a domain-specific one (biology)

Here is a picture of the editor. Any layout and styling change are welcome!

image

Regarding the pull request:

I've created a new view and an event file for the RAI extension on the editor codebase. I've added the new attributes to the metadata data class in core/state.py. Few more changes have been made to it.

On the other hand, the most critical part has been touching the Python library mlcommons to generate proper JSON to download. I basically touched the structure_graph/nodes/metadata.py, adding the attributes into the class declaration and the to_json() function. To properly use the export function, a new release of the ml-commons library needs to be released as a pip package. If not, I suspect it will generate problems.

Let's talk about it!

### CODE REVIEW

File: editor/core/constants.py [hunk 1]
```diff
@@ -34,7 +34,8 @@ def PAST_PROJECTS_PATH(user) -> epath.Path:
 METADATA = "Metadata"
 RESOURCES = "Resources"
 RECORD_SETS = "Record Sets"
-TABS = [OVERVIEW, METADATA, RESOURCES, RECORD_SETS]
+RAI = "Extension: Responsible AI"
+TABS = [OVERVIEW, METADATA, RESOURCES, RECORD_SETS, RAI]
```
Comment (marcenacp) [inline editor/core/constants.py:38]: Not sure where the "Responsible AI" tab should go:

- at the end?
- in-between "Metadata" and "Resources"?

File: python/mlcroissant/mlcroissant/_src/core/constants.py [hunk 2]
```diff
@@ -39,6 +39,7 @@
 ML_COMMONS_SUB_FIELD = ML_COMMONS.subField
 ML_COMMONS_SUB_FIELD_TYPE = ML_COMMONS.SubField
 ML_COMMONS_TRANSFORM = ML_COMMONS.transform
+ML_COMMONS_OTHER_FIELD = ML_COMMONS.otherField
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/core/constants.py:42]: Are RAI fields supposed to be nested in `otherField`?

File: python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py [hunk 3]
```diff
@@ -146,15 +166,34 @@ def to_json(self) -> Json:
...
+            "dataAnnotationPlatform": self.data_annotation_platform,
+            "dataAnnotationAnalysis": self.data_annotation_analysis,
+            "dataAnnotationPerItem": self.data_annotation_peritem,
+            "dataAnnotationDemographics": self.data_annotation_demographics,
+            "dataBiases": self.data_biases,
+            "dataUseCases": self.data_usecases,
+            "dataLimitations": self.data_limitation,
+            "dataSocialImpact": self.data_social_impact,
+            "dataSensitive": self.data_sensitive,
+            "dataMaitenance": self.data_maintenance,
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py:196]: dataMaintenance

File: python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py [hunk 4]
```diff
@@ -146,15 +166,34 @@ def to_json(self) -> Json:
...
             "distribution": [f.to_json() for f in self.distribution],
             "recordSet": [record_set.to_json() for record_set in self.record_sets],
+            ## RAI extension
+            "dataCollection": self.data_collection,
+            "dataCollectionType": self.data_collection_type,
+            "dataCollectionTypeOthers": self.data_collection_type_others,
+            "dataCollectionMissing": self.data_collection_missing,
+            "dataCollectionRaw": self.data_collection_raw,
+            "dataCollectionTimeFrameStart": data_collection_timeframe_start,
+            "dataCollectionTimeFrameEnd": data_collection_timeframe_end,<
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py:182]: Extra `<` at the end of the line.

File: python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py [hunk 5]
```diff
@@ -130,6 +144,12 @@ def to_json(self) -> Json:
         date_published = (
             self.date_published.isoformat() if self.date_published else None
         )
+        data_collection_timeframe_start = (
+            self.data_collection_timeframe_start.isoformat() if self.data_collection_timeframe_start else None
```
Comment (marcenacp) [inline python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py:148]: You should also update methods `Metadata.from_jsonld`/`Metadata.to_json` that respectively serialize/deserialize the JSON-LD

**Review by marcenacp — CHANGES_REQUESTED** [review]
Thanks for opening the PR! I left a few comments. It seems we still have to agree on some design ideas (which field describe RAI in JSON-LD, place of the new tab, etc).

Also:

- Can you sign the CLA in order for the GitHub Actions to trigger?
- Should you write some tests? The most common way for end-to-end tests is to have one of the datasets (https://github.com/mlcommons/croissant/tree/main/datasets/1.0) support RAI fields.

Thanks, Joan!

File: python/mlcroissant/mlcroissant/_src/core/constants.py (hunk 2 above)
Comment (JoanGi) [reply python/mlcroissant/mlcroissant/_src/core/constants.py:42]: No, it's a mistake. But! I think RAI attributes should be grouped somehow, as they are an extension of the core vocabulary. For instance, if someone adds another extension (e.g., domain-specific for biology). How can the new attributes be grouped in the spec?  I guess the easy way is to provide a wrapper like:

`  "ReponsibleAI": 
    {
"dataCollection": "Text",
"dataCollectionType": "Text",
"dataCollectionTypeOthers": "Text",
 "dataCollectionMissingData": "Text",
[...]
    },`

And then other extensions should follow the same

`  "CellBiology": 
    {
      "attribute1": "...",
      "attribute2": "...",
     [...]
    },`

What do you think? If not, we can set it as plain fields, as the example field you added few days ago.

File: python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py (hunk 5 above)
Comment (JoanGi) [reply python/mlcroissant/mlcroissant/_src/structure_graph/nodes/metadata.py:148]: As you pointed out earlier, the way to test the deserialization of the JSON-LD would be to update some dataset examples. However, example datasets are not properly loading, as I pointed out here: 
https://github.com/mlcommons/croissant/issues/478

**Comment by JoanGi** [comment]
recheck

**Comment by JoanGi** [comment]
recheck

**Comment by JoanGi** [comment]
recheck

**Comment by JoanGi** [comment]
Finally, I've managed to properly deserialize the JSON-LD into the editor for RAI attributes. Now, they can be created and loaded using the widget. We plan to add/modify some existing example datasets, adding proper RAI information before the launch.  

I've synced the fork because there were conflicting changes in the constant.py, and the PR has become a bit messy. For clarity, we can close this one, reopen a new one, remaking the changes. What do you think?

**Comment by JoanGi** [comment]
Closing PR, new PR opened more easy to follow and review at https://github.com/mlcommons/croissant/pull/487

**Pull request closed without merging**